pkg/k8s/allowlist_test.go

```diff
@@ -12,38 +12,159 @@
 	tests := []struct {
 		name           string
-		kind           string
+		gvk            k8sschema.GroupVersionKind
 		expectedResult bool
 	}{
 		{
-			name:           "ComputeInstance should support 'state-into-spec: merge'",
-			kind:           "ComputeInstance",
+			name: "ComputeInstance should support 'state-into-spec: merge'",
+			gvk: k8sschema.GroupVersionKind{
+				Group:   "compute.cnrm.cloud.google.com",
+				Version: "v1beta1",
+				Kind:    "ComputeInstance",
+			},
 			expectedResult: true,
 		},
 		{
 			name: "AccessContextManagerServicePerimeterResource should not " +
 				"support 'state-into-spec: merge'",
-			kind:           "AccessContextManagerServicePerimeterResource",
+			gvk: k8sschema.GroupVersionKind{
+				Group:   "accesscontextmanager.cnrm.cloud.google.com",
+				Version: "v1beta1",
+				Kind:    "AccessContextManagerServicePerimeterResource",
+			},
 			expectedResult: false,
 		},
 	}
 
 	for _, tc := range tests {
 		tc := tc
 		t.Run(tc.name, func(t *testing.T) {
-			actualResult := SupportsStateIntoSpecMergeInKind(tc.kind)
+			actualResult := k8s.SupportsStateIntoSpecMerge(tc.gvk)
 			if actualResult != tc.expectedResult {
 				t.Fatalf("got %v, want %v", actualResult, tc.expectedResult)
 			}
 		})
 	}
 }
+
+func TestOutputOnlyFieldsAreUnderObservedState(t *testing.T) {
+	crds, err := crdloader.LoadCRDs()
+	if err != nil {
+		t.Fatalf("error loading crds: %v", err)
+	}
+	smLoader, err := servicemappingloader.New()
+	if err != nil {
+		t.Fatalf("error getting new service mapping loader: %v", err)
+	}
+	for _, crd := range crds {
+		for _, version := range crd.Spec.Versions {
+			gvk := k8sschema.GroupVersionKind{
+				Group:   crd.Spec.Group,
+				Version: version.Name,
+				Kind:    crd.Spec.Names.Kind,
+			}
+			t.Run(fmt.Sprintf("%s/%s/%s", gvk.Group, gvk.Version, gvk.Kind), func(t *testing.T) {
+				openAPISchema := version.Schema.OpenAPIV3Schema
+				prop := findOpenAPIProperty(openAPISchema, "status", "observedState")
+				hasObservedState := prop != nil
+
+				// 'status.observedState' should exist for CRDs that
+				// (1) have all the output-only fields under 'status.observedState'
+				//     instead of 'status', or
+				// (2) have observedFields configured in the resource config to
+				//     expose selected output 'spec' fields.
+				// 'status.observedState' does not exist for CRDs that
+				// (1) don't have any output-only fields, or
+				// (2) have all the output-only fields under 'status' but don't
+				//     have observedFields configured.
+				mayHaveObservedState := k8s.OutputOnlyFieldsAreUnderObservedState(gvk)
+				rcs, err := smLoader.GetResourceConfigs(gvk)
+				// Ignore not found error because there are handwritten and
+				// DCL-based resources.
+				if err != nil && !strings.Contains(err.Error(), "no mapping with name") {
+					t.Errorf("error getting resource config(s) for gvk %+v: %v", gvk, err)
+				}
+				shouldHaveObservedState := false
+				for _, rc := range rcs {
```

Comment:
Not introduced in this PR, but I'm confused by the idea that there might be more than one service mapping.

Reply:
We have several instances of multiple ResourceConfigs mapping to the same Kind. This is mostly because we merged the locational resources, e.g. [rc1](https://cnrm.git.corp.google.com/cnrm/+/refs/heads/archived/master/config/servicemappings/compute.yaml#1368), [rc2](https://cnrm.git.corp.google.com/cnrm/+/refs/heads/archived/master/config/servicemappings/compute.yaml#93). Also we had to do some merges for edge case resources like ComputeInstance: [rc1](https://cnrm.git.corp.google.com/cnrm/+/refs/heads/archived/master/config/servicemappings/compute.yaml#816), [rc2](https://cnrm.git.corp.google.com/cnrm/+/refs/heads/archived/master/config/servicemappings/compute.yaml#923).